README.md
# Doublecount [![crates.io badge](https://img.shields.io/crates/v/doublecount.svg)](https://crates.io/crates/doublecount) [![docs.rs badge](https://docs.rs/doublecount/badge.svg)](https://docs.rs/doublecount/) [![license badge](https://img.shields.io/github/license/kellpossible/doublecount)](https://github.com/kellpossible/doublecount/blob/master/LICENSE.txt) [![github action badge](https://github.com/kellpossible/doublecount/workflows/Rust/badge.svg)](https://github.com/kellpossible/doublecount/actions?query=workflow%3ARust)

A double entry accounting system/library for Rust.

This project is very much inspired by [beancount](http://furius.ca/beancount/),
however it currently presents a much simpler model. It has been designed to
embed within other applications for the purpose of running accounting
calculations.

Commodities within the system are represented using the primitives provided by
the [commodity](https://crates.io/crates/commodity) library, which is in turn
backed by [rust_decimal](https://crates.io/crates/rust_decimal).

This library is under active development, however it should already be usable
for some simple purposes. There's likely to be some API changes in the future to
allow transactions/actions to be streamed into the system, and also to support
parallel computations of transactions to allow large programs to efficiently
executed on multi-core computers.

**[Changelog](./CHANGELOG.md)**

## Optional Features

The following features can be enabled to provide extra functionality:

+ `serde-support`
  + Enables support for serialization/de-serialization via `serde`

## Usage

```rust
use doublecount::{
    AccountStatus, EditAccountStatus, Account, Program, Action,
    ProgramState, Transaction, TransactionElement, BalanceAssertion,
    ActionTypeValue,
};
use commodity::{CommodityType, Commodity};
use chrono::NaiveDate;
use std::rc::Rc;
use std::str::FromStr;

// create a commodity from a currency's iso4317 alphanumeric code
let aud = Rc::from(CommodityType::from_str("AUD", "Australian Dollar").unwrap());

// Create a couple of accounts
let account1 = Rc::from(Account::new_with_id(Some("Account 1"), aud.id, None));
let account2 = Rc::from(Account::new_with_id(Some("Account 2"), aud.id, None));

// create a new program state, with accounts starting Closed
let mut program_state = ProgramState::new(
    &vec![account1.clone(), account2.clone()],
    AccountStatus::Closed
);

// open account1
let open_account1 = EditAccountStatus::new(
    account1.id,
    AccountStatus::Open,
    NaiveDate::from_str("2020-01-01").unwrap(),
);

// open account2
let open_account2 = EditAccountStatus::new(
    account2.id,
    AccountStatus::Open,
    NaiveDate::from_str("2020-01-01").unwrap(),
);

// create a transaction to transfer some commodity
// from account1 to account2.
let transaction1 = Transaction::new(
    Some(String::from("Transaction 1")),
    NaiveDate::from_str("2020-01-02").unwrap(),
    vec![
        TransactionElement::new(
            account1.id,
            Some(Commodity::from_str("-2.52 AUD").unwrap()),
            None,
        ),
        TransactionElement::new(
            account2.id,
            Some(Commodity::from_str("2.52 AUD").unwrap()),
            None,
        ),
    ],
);

// create a balance assertion (that will cause the program to return an error
// if it fails), to check that the balance of account1 matches the expected
// value of -1.52 AUD at the start of the date of 2020-01-03
let balance_assertion1 = BalanceAssertion::new(
    account1.id,
    NaiveDate::from_str("2020-01-03").unwrap(),
    Commodity::from_str("-2.52 AUD").unwrap()
);

// create another transaction to transfer commodity from
// account2 to account1, using the simpler syntax.
let transaction2 =  Transaction::new_simple(
   Some("Transaction 2"),
   NaiveDate::from_str("2020-01-03").unwrap(),
   account2.id,
   account1.id,
   Commodity::from_str("1.0 AUD").unwrap(),
   None,
);

let balance_assertion2 = BalanceAssertion::new(
    account1.id,
    NaiveDate::from_str("2020-01-04").unwrap(),
    Commodity::from_str("-1.52 AUD").unwrap()
);

let balance_assertion3 = BalanceAssertion::new(
    account2.id,
    NaiveDate::from_str("2020-01-04").unwrap(),
    Commodity::from_str("1.52 AUD").unwrap()
);

let actions: Vec<Rc<ActionTypeValue>> = vec![
    Rc::new(open_account1.into()),
    Rc::new(open_account2.into()),
    Rc::new(transaction1.into()),
    Rc::new(balance_assertion1.into()),
    Rc::new(transaction2.into()),
    Rc::new(balance_assertion2.into()),
    Rc::new(balance_assertion3.into()),
];

// create a program from the actions
let program = Program::new(actions);

// run the program
program_state.execute_program(&program).unwrap();
```


CHANGELOG.md
# Changelog

## v0.8.2

+ Update `commodity` to version `0.4`.

## v0.8.1

+ Fix incomplete `Action` implementation for `BalanceAssertion`.

## v0.8.0

+ Refactor to properly support user defined `Action`s. This included making `Program` and a number of other types generic over `ActionTypeValueEnum` and `ActionTypeEnum` implementations. These changes shouldn't affect anyone using the provided set of `Actions`.
+ Introduced a new trait `ActionTypeFor` which has some functionality separated from `Action`. This is a breaking change for people implementing `Action` themselves.

## v0.7.1

+ Fix building for default features accidentally including some items only needed for `serde-support`.

## v0.7.0

+ Complete `serde` serialization for `Program`, and the various `Action` implementations. [#2](https://github.com/kellpossible/doublecount/issues/2).
+ Refactor `Program` to use an enum `ActionTypeValue` instead of dynamic trait dispatch over `Action`. Programs using custom actions need to create their own implementation of `ActionTypeValueEnum` to provide store their actions given to `Program`.

## v0.6.2

+ Fix changelog fo `v0.6.1`

## v0.6.1

+ Bump `rust_decimal` dependency up to using generic version `1` to address [#5](https://github.com/kellpossible/doublecount/issues/5).
+ Update `Account#new()`, `Transaction#new()` and `Transaction#new_simple()` to use `Into<String>` to address [#4](https://github.com/kellpossible/doublecount/issues/4).

## v0.6.0

+ Renamed argument in `sum_account_states()`, `sum_currency` to
  `sum_commodity_type_id` to better match the recent changes in `commodity` library.

## v0.5.0

+ Updated `commodity` library dependency to `v0.3.0`, renamed some types
  which were changed for that version
+ `AccountingError::Currency` renamed to `AccountingError::Commodity`


src/error.rs
use super::{AccountID, AccountStatus, FailedBalanceAssertion, Transaction};
use commodity::exchange_rate::ExchangeRateError;
use commodity::{Commodity, CommodityError, CommodityTypeID};
use thiserror::Error;

/// An error associated with functionality in the [accounting](./index.html) module.
///
/// TODO: add context for the error for where it occurred within the [Program](super::Program)
#[derive(Error, Debug)]
pub enum AccountingError {
    #[error("error relating to a commodity")]
    Commodity(#[from] CommodityError),
    #[error("error relating to exchange rates")]
    ExchangeRate(#[from] ExchangeRateError),
    #[error("invalid account status ({:?}) for account {}", .status, .account_id)]
    InvalidAccountStatus {
        account_id: AccountID,
        status: AccountStatus,
    },
    #[error("error parsing a date from string")]
    DateParseError(#[from] chrono::ParseError),
    #[error("invalid transaction {0:?} because {1}")]
    InvalidTransaction(Transaction, String),
    #[error("failed checksum, the sum of account values in the common commodity type ({0}) does not equal zero")]
    FailedCheckSum(Commodity),
    #[error("no exchange rate supplied, unable to convert commodity {0} to type {1}")]
    NoExchangeRateSupplied(Commodity, CommodityTypeID),
    #[error("the account state with the id {0} was requested but cannot be found")]
    MissingAccountState(AccountID),
    #[error("the balance assertion failed {0}")]
    BalanceAssertionFailed(FailedBalanceAssertion),
}


src/account.rs
use arrayvec::ArrayString;
use commodity::{Commodity, CommodityTypeID};
use nanoid::nanoid;
use rust_decimal::Decimal;
use std::rc::Rc;

#[cfg(feature = "serde-support")]
use serde::{Deserialize, Serialize};

/// The size in characters/bytes of the [Account](Account) id.
const ACCOUNT_ID_LENGTH: usize = 20;

/// The status of an [Account](Account) stored within an [AccountState](AccountState).
#[cfg_attr(feature = "serde-support", derive(Serialize, Deserialize))]
#[derive(Copy, Clone, Debug, PartialEq)]
pub enum AccountStatus {
    /// The account is open
    Open,
    /// The account is closed
    Closed,
}
/// The type to use for the id of [Account](Account)s.
pub type AccountID = ArrayString<[u8; ACCOUNT_ID_LENGTH]>;

/// A way to categorize [Account](Account)s.
pub type AccountCategory = String;

/// Details for an account, which holds a [Commodity](Commodity)
/// with a type of [CommodityType](commodity::CommodityType).
#[cfg_attr(feature = "serde-support", derive(Serialize, Deserialize))]
#[derive(Debug, Clone)]
pub struct Account {
    /// A unique identifier for this `Account`, currently generated using [nanoid](nanoid).
    pub id: AccountID,

    /// The name of this `Account`
    pub name: Option<String>,

    /// The id of the type of commodity to be stored in this account
    pub commodity_type_id: CommodityTypeID,

    /// The category that this account part of
    pub category: Option<AccountCategory>,
}

impl Account {
    /// Create a new account with an automatically generated id (using
    /// [nanoid](nanoid)) and add it to this program state (and create
    /// its associated [AccountState](AccountState)).
    pub fn new_with_id<S: Into<String>>(
        name: Option<S>,
        commodity_type_id: CommodityTypeID,
        category: Option<AccountCategory>,
    ) -> Account {
        let id_string: String = nanoid!(ACCOUNT_ID_LENGTH);
        Self::new(
            ArrayString::from(id_string.as_ref()).unwrap_or_else(|_| {
                panic!(
                    "generated id string {0} should fit within ACCOUNT_ID_LENGTH: {1}",
                    id_string, ACCOUNT_ID_LENGTH
                )
            }),
            name,
            commodity_type_id,
            category,
        )
    }

    /// Create a new account and add it to this program state (and create its associated
    /// [AccountState](AccountState)).
    pub fn new<S: Into<String>>(
        id: AccountID,
        name: Option<S>,
        commodity_type_id: CommodityTypeID,
        category: Option<AccountCategory>,
    ) -> Account {
        Account {
            id,
            name: name.map(|s| s.into()),
            commodity_type_id,
            category,
        }
    }
}

impl PartialEq for Account {
    fn eq(&self, other: &Account) -> bool {
        self.id == other.id
    }
}

/// Mutable state associated with an [Account](Account).
#[derive(Debug, Clone, PartialEq)]
pub struct AccountState {
    /// The [Account](Account) associated with this state
    pub account: Rc<Account>,

    /// The amount of the commodity currently stored in this account
    pub amount: Commodity,

    /// The status of this account (open/closed/etc...)
    pub status: AccountStatus,
}

impl AccountState {
    /// Create a new [AccountState](AccountState).
    pub fn new(account: Rc<Account>, amount: Commodity, status: AccountStatus) -> AccountState {
        AccountState {
            account,
            amount,
            status,
        }
    }

    /// Open this account, set the `status` to [Open](AccountStatus::Open)
    pub fn open(&mut self) {
        self.status = AccountStatus::Open;
    }

    // Close this account, set the `status` to [Closed](AccountStatus::Closed)
    pub fn close(&mut self) {
        self.status = AccountStatus::Closed;
    }

    pub fn eq_approx(&self, other: &AccountState, epsilon: Decimal) -> bool {
        self.account == other.account
            && self.status == other.status
            && self.amount.eq_approx(other.amount, epsilon)
    }
}

#[cfg(feature = "serde-support")]
#[cfg(test)]
mod serde_tests {
    use super::Account;
    use super::AccountID;
    use commodity::CommodityTypeID;
    use std::str::FromStr;

    #[test]
    fn account_serde() {
        use serde_json;

        let json = r#"{
  "id": "ABCDEFGHIJKLMNOPQRST",
  "name": "Test Account",
  "commodity_type_id": "USD",
  "category": "Expense"
}"#;

        let account: Account = serde_json::from_str(json).unwrap();

        let reference_account = Account::new(
            AccountID::from("ABCDEFGHIJKLMNOPQRST").unwrap(),
            Some("TestAccount"),
            CommodityTypeID::from_str("AUD").unwrap(),
            Some("Expense".to_string()),
        );

        assert_eq!(reference_account, account);
        insta::assert_json_snapshot!(account);
    }
}


src/program.rs
use super::{
    Account, AccountID, AccountState, AccountStatus, AccountingError, ActionOrder,
    FailedBalanceAssertion,
};
use commodity::exchange_rate::ExchangeRate;
use commodity::{Commodity, CommodityTypeID};
use std::collections::HashMap;
use std::marker::PhantomData;
use std::rc::Rc;

use crate::{ActionType, ActionTypeFor, ActionTypeValue, ActionTypeValueEnum};
#[cfg(feature = "serde-support")]
use serde::{de, ser::SerializeSeq, Deserialize, Deserializer, Serialize, Serializer};

/// A collection of [Action](Action)s to be executed in order to
/// mutate some [ProgramState](ProgramState).
#[derive(Debug, Clone, PartialEq)]
pub struct Program<AT = ActionType, ATV = ActionTypeValue> {
    pub actions: Vec<Rc<ATV>>,
    action_type: PhantomData<AT>,
}

impl<AT, ATV> Program<AT, ATV>
where
    AT: Ord,
    ATV: ActionTypeValueEnum<AT> + ActionTypeFor<AT>,
{
    /// Create a new [Program](Program).
    ///
    /// The provided `actions` will be sorted using [ActionOrder](ActionOrder).
    pub fn new(actions: Vec<Rc<ATV>>) -> Program<AT, ATV> {
        let mut sorted_actions: Vec<Rc<ATV>> = actions;
        sorted_actions.sort_by_key(|a| ActionOrder::new(a.clone()));
        Program {
            actions: sorted_actions,
            action_type: PhantomData::default(),
        }
    }

    /// The number of actions in this program.
    pub fn len(&self) -> usize {
        self.actions.len()
    }

    /// Returns true if there are no actions in this progam.
    pub fn is_empty(&self) -> bool {
        self.actions.is_empty()
    }
}

#[cfg(feature = "serde-support")]
struct ProgramVisitor<AT, ATV> {
    action_type: PhantomData<AT>,
    action_type_value: PhantomData<ATV>,
}

#[cfg(feature = "serde-support")]
impl<AT, ATV> ProgramVisitor<AT, ATV> {
    pub fn new() -> Self {
        Self {
            action_type: PhantomData::default(),
            action_type_value: PhantomData::default(),
        }
    }
}

#[cfg(feature = "serde-support")]
impl<'de, AT, ATV> de::Visitor<'de> for ProgramVisitor<AT, ATV>
where
    AT: Ord,
    ATV: Deserialize<'de> + ActionTypeValueEnum<AT> + ActionTypeFor<AT>,
{
    type Value = Program<AT, ATV>;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        formatter.write_str("Program comprising of a vector of Actions")
    }

    fn visit_seq<S>(self, mut seq: S) -> Result<Program<AT, ATV>, S::Error>
    where
        S: de::SeqAccess<'de>,
    {
        let mut actions: Vec<Rc<ATV>> = match seq.size_hint() {
            Some(size_hint) => Vec::with_capacity(size_hint),
            None => Vec::new(),
        };

        while let Some(action) = seq.next_element::<ATV>()? {
            actions.push(Rc::new(action));
        }

        Ok(Program::new(actions))
    }
}

#[cfg(feature = "serde-support")]
impl<'de, AT, ATV> Deserialize<'de> for Program<AT, ATV>
where
    AT: Ord,
    ATV: Deserialize<'de> + ActionTypeValueEnum<AT> + ActionTypeFor<AT>,
{
    fn deserialize<D>(deserializer: D) -> std::result::Result<Program<AT, ATV>, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_seq(ProgramVisitor::<AT, ATV>::new())
    }
}

#[cfg(feature = "serde-support")]
impl<AT, ATV> Serialize for Program<AT, ATV>
where
    ATV: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut seq = serializer.serialize_seq(Some(self.actions.len()))?;
        for e in &self.actions {
            seq.serialize_element(&**e)?;
        }
        seq.end()
    }
}

/// The state of a [Program](Program) being executed.
pub struct ProgramState<AT = ActionType, ATV = ActionTypeValue> {
    /// list of states associated with accounts (can only grow)
    pub account_states: HashMap<AccountID, AccountState>,

    /// list of failed assertions, and associated failed balance
    pub failed_balance_assertions: Vec<FailedBalanceAssertion>,

    /// the index of the currently executing action
    current_action_index: usize,

    action_type: PhantomData<AT>,
    action_type_value: PhantomData<ATV>,
}

/// Sum the values in all the accounts into a single
/// [Commodity](Commodity), and use the supplied exchange rate if
/// required to convert a type of commodity in an account to the
/// [CommidityType](commodity::CommodityType) associated with the
/// id `sum_commodity_type_id`.
pub fn sum_account_states(
    account_states: &HashMap<AccountID, AccountState>,
    sum_commodity_type_id: CommodityTypeID,
    exchange_rate: Option<&ExchangeRate>,
) -> Result<Commodity, AccountingError> {
    let mut sum = Commodity::zero(sum_commodity_type_id);

    for account_state in account_states.values() {
        let account_amount = if account_state.amount.type_id != sum_commodity_type_id {
            match exchange_rate {
                Some(rate) => rate.convert(account_state.amount, sum_commodity_type_id)?,
                None => {
                    return Err(AccountingError::NoExchangeRateSupplied(
                        account_state.amount,
                        sum_commodity_type_id,
                    ))
                }
            }
        } else {
            account_state.amount
        };

        sum = sum.add(&account_amount)?;
    }

    Ok(sum)
}

impl<AT, ATV> ProgramState<AT, ATV>
where
    ATV: ActionTypeValueEnum<AT>,
{
    /// Create a new [ProgramState](ProgramState).
    pub fn new(accounts: &[Rc<Account>], account_status: AccountStatus) -> ProgramState<AT, ATV> {
        let mut account_states = HashMap::new();

        for account in accounts {
            account_states.insert(
                account.id,
                AccountState::new(
                    account.clone(),
                    Commodity::zero(account.commodity_type_id),
                    account_status,
                ),
            );
        }

        ProgramState {
            account_states,
            failed_balance_assertions: Vec::new(),
            current_action_index: 0,
            action_type: PhantomData::default(),
            action_type_value: PhantomData::default(),
        }
    }

    /// Execute a given [Program](Program) to mutate this state.
    pub fn execute_program(&mut self, program: &Program<AT, ATV>) -> Result<(), AccountingError> {
        for (index, action) in program.actions.iter().enumerate() {
            action.as_action().perform(self)?;
            self.current_action_index = index;
        }

        // TODO: change this to return a list of failed assertions in the error
        if let Some(failed_assertion) = self.failed_balance_assertions.get(0) {
            return Err(AccountingError::BalanceAssertionFailed(
                failed_assertion.clone(),
            ));
        }

        Ok(())
    }

    /// Get the reference to an [Account](Account) using it's [AccountID](AccountID).
    pub fn get_account(&self, account_id: &AccountID) -> Option<&Account> {
        self.get_account_state(account_id)
            .map(|state| state.account.as_ref())
    }

    /// Get a reference to the `AccountState` associated with a given `Account`.
    pub fn get_account_state(&self, account_id: &AccountID) -> Option<&AccountState> {
        self.account_states.get(account_id)
    }

    /// Get a mutable reference to the `AccountState` associated with the given `Account`.
    pub fn get_account_state_mut(&mut self, account_id: &AccountID) -> Option<&mut AccountState> {
        self.account_states.get_mut(account_id)
    }

    /// Record a failed [BalanceAssertion](super::BalanceAssertion)
    /// using a [FailedBalanceAssertion](FailedBalanceAssertion).
    pub fn record_failed_balance_assertion(
        &mut self,
        failed_balance_assertion: FailedBalanceAssertion,
    ) {
        self.failed_balance_assertions
            .push(failed_balance_assertion);
    }
}

#[cfg(feature = "serde-support")]
#[cfg(test)]
mod tests {
    use super::Program;
    use crate::{
        Account, AccountID, AccountStatus, ActionTypeValue, BalanceAssertion, EditAccountStatus,
        Transaction, TransactionElement,
    };
    use chrono::NaiveDate;
    use commodity::{Commodity, CommodityType, CommodityTypeID};
    use std::{rc::Rc, str::FromStr};

    #[test]
    fn program_serde() {
        let json = r#"
[
    {
        "type": "EditAccountStatus",
        "account_id": "TestAccount1",
        "newstatus": "Open",
        "date": "2020-01-01"
    },
    {
        "type": "EditAccountStatus",
        "account_id": "TestAccount2",
        "newstatus": "Open",
        "date": "2020-01-01"
    },
    {
        "type": "Transaction",
        "description": "Test Transaction",
        "date": "2020-01-02",
        "elements": [
            {
                "account_id": "TestAccount1",
                "amount": {
                    "value": "-2.52",
                    "type_id": "AUD"
                }
            },
            {
                "account_id": "TestAccount2",
                "amount": {
                    "value": "2.52",
                    "type_id": "AUD"
                }
            }
        ]  
    },
    {
        "type": "BalanceAssertion",
        "account_id": "TestAccount1",
        "date": "2020-01-03",
        "expected_balance": {
            "value": "-3.52",
            "type_id": "AUD"
        }
    }
]"#;
        let program: Program = serde_json::from_str(json).unwrap();

        let aud = Rc::from(CommodityType::new(
            CommodityTypeID::from_str("AUD").unwrap(),
            None,
        ));

        let account1 = Rc::from(Account::new(
            AccountID::from("TestAccount1").unwrap(),
            Some("Test Account 1"),
            aud.id,
            None,
        ));
        let account2 = Rc::from(Account::new(
            AccountID::from("TestAccount2").unwrap(),
            Some("Test Account 2"),
            aud.id,
            None,
        ));

        let open_account1 = EditAccountStatus::new(
            account1.id,
            AccountStatus::Open,
            NaiveDate::from_str("2020-01-01").unwrap(),
        );

        let open_account2 = EditAccountStatus::new(
            account2.id,
            AccountStatus::Open,
            NaiveDate::from_str("2020-01-01").unwrap(),
        );

        let transaction = Transaction::new(
            Some(String::from("Test Transaction")),
            NaiveDate::from_str("2020-01-02").unwrap(),
            vec![
                TransactionElement::new(
                    account1.id,
                    Some(Commodity::from_str("-2.52 AUD").unwrap()),
                    None,
                ),
                TransactionElement::new(
                    account2.id,
                    Some(Commodity::from_str("2.52 AUD").unwrap()),
                    None,
                ),
            ],
        );

        let balance_assertion = BalanceAssertion::new(
            account1.id,
            NaiveDate::from_str("2020-01-03").unwrap(),
            Commodity::from_str("-3.52 AUD").unwrap(),
        );

        let actions: Vec<Rc<ActionTypeValue>> = vec![
            Rc::new(open_account1.into()),
            Rc::new(open_account2.into()),
            Rc::new(transaction.into()),
            Rc::new(balance_assertion.into()),
        ];

        let reference_program = Program::new(actions);

        assert_eq!(reference_program, program);

        insta::assert_json_snapshot!(program);
    }
}


src/lib.rs
//! A double entry accounting system/library.
//!
//! # Optional Features
//!
//! The doublecount package has the following optional cargo features:
//!
//! + `serde-support`
//!   + Disabled by default
//!   + Enables support for serialization/de-serialization via `serde`
//!
//! # Usage
//!
//! ```
//! use doublecount::{
//!     AccountStatus, EditAccountStatus, Account, Program, Action,
//!     ProgramState, Transaction, TransactionElement, BalanceAssertion,
//!     ActionTypeValue,
//! };
//! use commodity::{CommodityType, Commodity};
//! use chrono::NaiveDate;
//! use std::rc::Rc;
//! use std::str::FromStr;
//!
//! // create a commodity from a currency's iso4317 alphanumeric code
//! let aud = Rc::from(CommodityType::from_str("AUD", "Australian Dollar").unwrap());
//!
//! // Create a couple of accounts
//! let account1 = Rc::from(Account::new_with_id(Some("Account 1"), aud.id, None));
//! let account2 = Rc::from(Account::new_with_id(Some("Account 2"), aud.id, None));
//!
//! // create a new program state, with accounts starting Closed
//! let mut program_state = ProgramState::new(
//!     &vec![account1.clone(), account2.clone()],
//!     AccountStatus::Closed
//! );
//!
//! // open account1
//! let open_account1 = EditAccountStatus::new(
//!     account1.id,
//!     AccountStatus::Open,
//!     NaiveDate::from_str("2020-01-01").unwrap(),
//! );
//!
//! // open account2
//! let open_account2 = EditAccountStatus::new(
//!     account2.id,
//!     AccountStatus::Open,
//!     NaiveDate::from_str("2020-01-01").unwrap(),
//! );
//!
//! // create a transaction to transfer some commodity
//! // from account1 to account2.
//! let transaction1 = Transaction::new(
//!     Some(String::from("Transaction 1")),
//!     NaiveDate::from_str("2020-01-02").unwrap(),
//!     vec![
//!         TransactionElement::new(
//!             account1.id,
//!             Some(Commodity::from_str("-2.52 AUD").unwrap()),
//!             None,
//!         ),
//!         TransactionElement::new(
//!             account2.id,
//!             Some(Commodity::from_str("2.52 AUD").unwrap()),
//!             None,
//!         ),
//!     ],
//! );
//!
//! // create a balance assertion (that will cause the program to return an error
//! // if it fails), to check that the balance of account1 matches the expected
//! // value of -1.52 AUD at the start of the date of 2020-01-03
//! let balance_assertion1 = BalanceAssertion::new(
//!     account1.id,
//!     NaiveDate::from_str("2020-01-03").unwrap(),
//!     Commodity::from_str("-2.52 AUD").unwrap()
//! );
//!
//! // create another transaction to transfer commodity from
//! // account2 to account1, using the simpler syntax.
//! let transaction2 =  Transaction::new_simple(
//!    Some("Transaction 2"),
//!    NaiveDate::from_str("2020-01-03").unwrap(),
//!    account2.id,
//!    account1.id,
//!    Commodity::from_str("1.0 AUD").unwrap(),
//!    None,
//! );
//!
//! let balance_assertion2 = BalanceAssertion::new(
//!     account1.id,
//!     NaiveDate::from_str("2020-01-04").unwrap(),
//!     Commodity::from_str("-1.52 AUD").unwrap()
//! );
//!
//! let balance_assertion3 = BalanceAssertion::new(
//!     account2.id,
//!     NaiveDate::from_str("2020-01-04").unwrap(),
//!     Commodity::from_str("1.52 AUD").unwrap()
//! );
//!
//! let actions: Vec<Rc<ActionTypeValue>> = vec![
//!     Rc::new(open_account1.into()),
//!     Rc::new(open_account2.into()),
//!     Rc::new(transaction1.into()),
//!     Rc::new(balance_assertion1.into()),
//!     Rc::new(transaction2.into()),
//!     Rc::new(balance_assertion2.into()),
//!     Rc::new(balance_assertion3.into()),
//! ];
//!
//! // create a program from the actions
//! let program = Program::new(actions);
//!
//! // run the program
//! program_state.execute_program(&program).unwrap();
//! ```

extern crate arrayvec;
extern crate chrono;
extern crate commodity;
extern crate nanoid;
extern crate rust_decimal;
extern crate thiserror;

#[cfg(feature = "serde-support")]
extern crate serde;

#[cfg(test)]
#[cfg(feature = "serde-support")]
extern crate serde_json;

mod account;
mod actions;
mod error;
mod program;

pub use account::*;
pub use actions::*;
pub use error::AccountingError;
pub use program::*;

#[cfg(doctest)]
#[macro_use]
extern crate doc_comment;

#[cfg(doctest)]
doctest!("../README.md");

#[cfg(test)]
mod tests {
    use super::{
        sum_account_states, Account, AccountState, AccountStatus, BalanceAssertion,
        EditAccountStatus, Program, ProgramState, Transaction, TransactionElement,
    };
    use crate::ActionTypeValue;
    use chrono::NaiveDate;
    use commodity::{Commodity, CommodityType, CommodityTypeID};
    use std::rc::Rc;
    use std::str::FromStr;

    #[test]
    fn execute_program() {
        let aud = Rc::from(CommodityType::new(
            CommodityTypeID::from_str("AUD").unwrap(),
            None,
        ));
        let account1 = Rc::from(Account::new_with_id(Some("Account 1"), aud.id, None));
        let account2 = Rc::from(Account::new_with_id(Some("Account 2"), aud.id, None));

        let accounts = vec![account1.clone(), account2.clone()];

        let mut program_state = ProgramState::new(&accounts, AccountStatus::Closed);

        let open_account1 = EditAccountStatus::new(
            account1.id,
            AccountStatus::Open,
            NaiveDate::from_str("2020-01-01").unwrap(),
        );

        let open_account2 = EditAccountStatus::new(
            account2.id,
            AccountStatus::Open,
            NaiveDate::from_str("2020-01-01").unwrap(),
        );

        let transaction1 = Transaction::new(
            Some(String::from("Transaction 1")),
            NaiveDate::from_str("2020-01-02").unwrap(),
            vec![
                TransactionElement::new(
                    account1.id,
                    Some(Commodity::from_str("-2.52 AUD").unwrap()),
                    None,
                ),
                TransactionElement::new(
                    account2.id,
                    Some(Commodity::from_str("2.52 AUD").unwrap()),
                    None,
                ),
            ],
        );

        let transaction2 = Transaction::new(
            Some(String::from("Transaction 2")),
            NaiveDate::from_str("2020-01-02").unwrap(),
            vec![
                TransactionElement::new(
                    account1.id,
                    Some(Commodity::from_str("-1.0 AUD").unwrap()),
                    None,
                ),
                TransactionElement::new(account2.id, None, None),
            ],
        );

        let balance_assertion = BalanceAssertion::new(
            account1.id,
            NaiveDate::from_str("2020-01-03").unwrap(),
            Commodity::from_str("-3.52 AUD").unwrap(),
        );

        let actions: Vec<Rc<ActionTypeValue>> = vec![
            Rc::new(open_account1.into()),
            Rc::new(open_account2.into()),
            Rc::new(transaction1.into()),
            Rc::new(transaction2.into()),
            Rc::new(balance_assertion.into()),
        ];

        let program = Program::new(actions);

        let account1_state_before: AccountState = program_state
            .get_account_state(&account1.id)
            .unwrap()
            .clone();

        assert_eq!(AccountStatus::Closed, account1_state_before.status);

        program_state.execute_program(&program).unwrap();

        let account1_state_after: AccountState = program_state
            .get_account_state(&account1.id)
            .unwrap()
            .clone();

        assert_eq!(AccountStatus::Open, account1_state_after.status);
        assert_eq!(
            Commodity::from_str("-3.52 AUD").unwrap(),
            account1_state_after.amount
        );

        assert_eq!(
            Commodity::from_str("0.0 AUD").unwrap(),
            sum_account_states(
                &program_state.account_states,
                CommodityTypeID::from_str("AUD").unwrap(),
                None
            )
            .unwrap()
        );
    }
}
